### PR TITLE
FS-51 Report unavailable clients for Proteus messages

### PR DESCRIPTION
**Pull request by lepsa**
Changing the return types to match the ticket.
Adding tests and fixing some logic errors.

Sending a Proteus message will no longer emit an error when the conversation owning server cannot contact federated servers to retrieve client lists. These unavailable users will be listed in the failed to send field of the response sent back to the client. Federation members that can be contacted will have messages fanned out to them.

## Checklist

 - [:heavy_check_mark:] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [:heavy_check_mark:] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: services/galley/test/integration/API.hs
Comment (mdimjasevic) [inline services/galley/test/integration/API.hs:1220]: Why is `deeClient` a redundant recipient?

File: services/galley/src/Galley/API/Message.hs
Comment (mdimjasevic) [inline services/galley/src/Galley/API/Message.hs:513]: This doesn't seem so straightforward to understand, but perhaps it was the easiest way to write it.

File: services/galley/src/Galley/API/Message.hs
Comment (mdimjasevic) [inline services/galley/src/Galley/API/Message.hs:500]: Can you please write a few tests that check various cases that can happen here? I don't think the current test provides a good coverage wrt this part of the code.

**Review by mdimjasevic — COMMENTED** [review]
Overall, this looks good, but there are some parts that are not that easy for me to grasp. It'd be great if we could go over the PR in a call!

**Comment by mdimjasevic** [comment]
@lepsa , please take a look at the federation chat and the reasoning why this is a breaking API change. I'd be happy to pair with you tomorrow morning (Wednesday) should you need help with splitting an endpoint in `wire-api` into two, one for versions up to and including v2, and a new one for all versions from v3 and up. I also tried to incorporate your structure suggestion in Confluence: https://wearezeta.atlassian.net/wiki/spaces/CORE/pages/738557958/Offline+backend+handling+-+analysis+of+prekey+fetching+Proteus#Suggested-change

File: services/galley/test/integration/API.hs
Comment (lepsa) [reply services/galley/test/integration/API.hs:1220]: It is due to other tests expecting redundant clients under similar circumstances. If I add code to filter out users from the redundant field those tests will fail.

File: services/galley/src/Galley/API/Message.hs
Comment (lepsa) [reply services/galley/src/Galley/API/Message.hs:513]: It basically is the easiest way. It moves into each of the Map values by the given keys and appends the client sets. It needs to be `alter` however so that new keys can be added, which requires the update function to be `Maybe a -> Maybe a`, hence all the `pure`s and `fromMaybe`s.

I'll update the names though.

File: services/galley/test/integration/API.hs
Comment (mdimjasevic) [reply services/galley/test/integration/API.hs:1220]: Ok, but I still don't understand why. It would also be beneficial for your understanding to get to the bottom of it.

File: services/galley/test/integration/API.hs
Comment (mdimjasevic) [inline services/galley/test/integration/API.hs:1252]: The tests in here are unit tests, aren't they? Can you move them to a unit test module?

**Review by mdimjasevic — APPROVED** [review]
Apart from moving unit tests into a unit test module, this looks good!

**Comment by mdimjasevic** [comment]
Please move the unit tests and then merge. No need to wait for another approval from me.

**Pull request merged**
